src/System.Memory/ref/System.Memory.cs

```diff
@@ -112,6 +112,21 @@ public static partial class MemoryExtensions
 }
 namespace System.Buffers
 {
+    public sealed partial class ArrayBufferWriter<T> : System.Buffers.IBufferWriter<T>, System.IDisposable
+    {
+        public ArrayBufferWriter() { }
+        public ArrayBufferWriter(int initialCapacity) { }
+        public int AvailableSpace { get { throw null; } }
+        public int Capacity { get { throw null; } }
+        public int CurrentIndex { get { throw null; } }
+        public System.ReadOnlyMemory<T> OutputAsMemory { get { throw null; } }
+        public System.ReadOnlySpan<T> OutputAsSpan { get { throw null; } }
```

Comment:
Do we need to expose both OutputAsMemory and OutputAsSpan?  You can get a Span from a Memory, so the latter is equivalent to OutputAsMemory.Span.  Given the typical use for this type, does the difference in perf between the two when you actually want a span make any measurable difference?

Same question goes for exposing GetSpan; as long as GetMemory is exposed, GetSpan could be implemented explicitly.

Reply:
Given my primary scenario for this was writing the data to a stream (asynchronously), we can defer adding `OutputAsSpan` until there is a customer request for it (Mem.Span is not noticeably slower (i.e. 15-25 ns) even for small writes like hello world since it's called infrequently).

> Same question goes for exposing GetSpan; as long as GetMemory is exposed, GetSpan could be implemented explicitly.

However, I don't know if the same applies to GetSpan/GetMemory. If we expect `ArrayBufferWriter` to be used similar to how `IBufferWriter` is used today (i.e. we have Advance/GetMemory as regular interface implementations), we should be consistent and implement GetSpan that way as well. Writing some data to a span (synchronously) is faster than going through memory (especially if its a small amount). That was the primary reason to have `GetSpan` on the interface to begin with. I haven't measured myself but presumably the perf was significant enough to justify adding it, given how we currently use the API. It looks like we do this often with high granularity:
https://github.com/aspnet/AspNetCore/blob/68cc5e7846240bacc146854dfc6a175e30ff0f60/src/Servers/Kestrel/Core/src/Internal/BufferWriter.cs#L20
https://github.com/aspnet/AspNetCore/blob/b1f778bfb887f5c8282f5de42302c3f5736da790/src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs#L404
https://github.com/aspnet/AspNetCore/blob/3cd5054eb5d8851b25881cd05cccb931aaa022e6/src/SignalR/common/Shared/TextMessageFormatter.cs#L17

We only call GetMemory for async calls or interop where the work being done is expensive already:
https://github.com/aspnet/AspNetCore/blob/6d442c1e3d7875132c662a9e8c537e30cea585b8/src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs#L119
https://github.com/aspnet/AspNetCore/blob/68cc5e7846240bacc146854dfc6a175e30ff0f60/src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvConnection.cs#L148
https://github.com/aspnet/AspNetCore/blob/ad11f890ef0646ecf3cefde63c9021865aa9490e/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L165
https://github.com/aspnet/AspNetCore/blob/3cd5054eb5d8851b25881cd05cccb931aaa022e6/src/SignalR/clients/csharp/Http.Connections.Client/src/Internal/WebSocketsTransport.cs#L222

Reply:
A benchmark like this would regress ~5%:
https://github.com/aspnet/AspNetCore/blob/dbf82dc8c4cdb410f036811143812db9f8bcc096/src/SignalR/perf/Microbenchmarks/HandshakeProtocolBenchmark.cs#L81
https://github.com/aspnet/AspNetCore/blob/dbf82dc8c4cdb410f036811143812db9f8bcc096/src/SignalR/common/SignalR.Common/src/Protocol/HandshakeProtocol.cs#L59-L70

``` ini

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.253 (1809/October2018Update/Redstone5)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview-010174
  [Host] : .NET Core 3.0.0-preview-27324-5 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT
  Core   : .NET Core 3.0.0-preview-27324-5 (CoreCLR 4.6.27322.0, CoreFX 4.7.19.7311), 64bit RyuJIT

Job=Core  EnvironmentVariables=COMPlus_TieredCompilation=0  Runtime=Core  
Toolchain=.NET Core 3.0  

```
|                        Method |     Mean |    Error |   StdDev | Ratio | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------------------------ |---------:|---------:|---------:|------:|------------:|------------:|------------:|--------------------:|
| WriteArrayBufferWriterMemSpan | 257.4 ns | 2.561 ns | 2.396 ns |  1.05 |      0.0076 |           - |           - |                32 B |
|     WriteArrayBufferWriterSpan | 245.8 ns | 2.107 ns | 1.645 ns |  1.00 |      0.0076 |           - |           - |                32 B |

Reply:
😢